common/lib/xmodule/xmodule/modulestore/__init__.py

```diff
@@ -270,14 +270,22 @@ def _end_bulk_operation(self, structure_key, emit_signals=True):
         if bulk_ops_record.active:
             return
 
-        dirty = self._end_outermost_bulk_operation(bulk_ops_record, structure_key)
+        # The bulk op has ended. However, the signal tasks below still need to use the
+        # built-up bulk op information (if the signals trigger tasks in the same thread).
```

Comment:
You're correct that production and non-production would behave differently, since production delegates signal-triggered tasks to celery and therefore a different thread. In a non-production environment however, the code before your recent PR appeared to be using the already-built-up bulk op when executing the signal-triggered tasks on the same thread. This change attempts to preserve that behavior. In my local devstack tests, this PR's change decreased the duration of a course import from ~130 seconds to ~40 seconds.

I agree that the task you mentioned should be wrapped in a bulk op. And yes - it would have been masked before. Generally, anywhere that `modulestore().get_course()` is called - or really most any modulestore call, the code should be wrapped in a bulk op.